implement putmask function

rushabh-v opened the pull request:
I just found one more important function which I accidentally missed to add in my GSoC proposal, So, I have implemented it here.

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +69,70 @@ def _can_memcpy(dst, src):
+        raise ValueError('mask and data must be the same size')
+
+    if a.shape == values.shape:
+        putmask_kernel(mask.astype(a.dtype), values, a)
```
```suggestion
        a[mask] = value
```

rushabh-v replied:
Oh, yes that's quite better.

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +69,70 @@ def _can_memcpy(dst, src):
+        len_vals = len(values)
+        n = 0
+
+        for i in range(len(a)):
```
Please parallelize this loop.

rushabh-v replied:
I think this loop can not be parallelized because decisions to be made in every iteration depends on previous iterations. I tried but couldn't figure out a way to make an indexed mask without looping. Do you have any other plans?

asi1024 replied:
`n` can be calculated by `cupy.cumsum(mask)`

asi1024 commented on tests/cupy_tests/manipulation_tests/test_basic.py:
```diff
@@ -125,3 +126,44 @@ def test_copyto_where(self, xp, dtype):
+@testing.gpu
+class TestPutmask(unittest.TestCase):
+
+    @testing.for_all_dtypes(no_complex=True, no_bool=True)
```
```suggestion
    @testing.for_all_dtypes()
```

rushabh-v replied:
The operation(power) used for the array `values` in this test can not be performed on `bool`s. That's why I am going to write a separate test for `bool`s.

asi1024 replied:
So, do not use power and modulo in this test.

asi1024 commented on tests/cupy_tests/manipulation_tests/test_basic.py:
```diff
@@ -125,3 +126,44 @@ def test_copyto_where(self, xp, dtype):
+
+    def test_putmask(self):
+        for xp in (numpy, cupy):
+            a = cupy.array([1, 2, 3])
```
```suggestion
            a = xp.array([1, 2, 3])
```

rushabh-v replied:
Oh, I missed it...

asi1024 commented on tests/cupy_tests/manipulation_tests/test_basic.py:
```diff
@@ -125,3 +126,44 @@ def test_copyto_where(self, xp, dtype):
+@testing.gpu
+class TestPutmaskNonEqual(unittest.TestCase):
+
+    @testing.for_all_dtypes(no_complex=True, no_bool=True)
```
```suggestion
    @testing.for_all_dtypes()
```

rushabh-v replied:
ditto

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -116,23 +116,21 @@ def putmask(a, mask, values):
-                a[i] = values[n % len_vals]
-                n += 1
-
+        mask = cupy.cumsum(mask.astype(int))
```
```suggestion
        mask = cupy.cumsum(mask, dtype=numpy.int64)
```

rushabh-v commented:
I have parallelize that loop, @asi1024 . We don't need to use even `cumsum` here. Actually I misunderstood the numpy's functionality before because of its confusing example in their docstring. I have also updated our docstring's example to not be confusing.

asi1024 commented on tests/cupy_tests/manipulation_tests/test_basic.py:
```diff
@@ -125,3 +126,47 @@ def test_copyto_where(self, xp, dtype):
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_allclose()
+    def test_putmask(self, xp, dtype):
+        a = testing.shaped_arange(self.shape, xp, dtype=dtype)
+        mask = testing.shaped_random(self.shape, xp, dtype=bool)
+        values = testing.shaped_random(self.shape, xp, dtype=dtype)
```
Feed different seeds.
```suggestion
        a = testing.shaped_arange(self.shape, xp, dtype=dtype, seed=0)
        mask = testing.shaped_random(self.shape, xp, dtype=bool, seed=1)
        values = testing.shaped_random(self.shape, xp, dtype=dtype, seed=2)
```

asi1024 commented on tests/cupy_tests/manipulation_tests/test_basic.py:
```diff
@@ -125,3 +126,47 @@ def test_copyto_where(self, xp, dtype):
+        a = testing.shaped_arange(self.shape, xp, dtype=dtype)
+        mask = testing.shaped_random(self.shape, xp, dtype=bool)
+        values = testing.shaped_random(self.shape, xp, dtype=dtype)
+        xp.putmask(a, mask, values)
```
```suggestion
        ret = xp.putmask(a, mask, values)
        assert ret is None
```

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +69,61 @@ def _can_memcpy(dst, src):
+        if values.ndim > 1:
+            values = values.flatten()
+
+        putmask_kernel(mask.astype(int), values, len(values), a)
```
Please use numpy dtype for the readability.
```suggestion
        putmask_kernel(mask.astype(numpy.int64), values, len(values), a)
```

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +69,61 @@ def _can_memcpy(dst, src):
+putmask_kernel = core._kernel.ElementwiseKernel(
+    'Q mask, raw T values, uint64 len_vals', 'T out',
+    '''
+    if (mask) out = values[(i % len_vals)];
```
```suggestion
    if (mask) out = values[i % len_vals];
```

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +69,61 @@ def _can_memcpy(dst, src):
         dst.size == src.size
+
+
+putmask_kernel = core._kernel.ElementwiseKernel(
```
Please add single underscore prefix as it is private function.
```suggestion
_putmask_kernel = core._kernel.ElementwiseKernel(
```

asi1024 commented:
Please add tests for the combination of dtypes of `a` and `value`.

rushabh-v commented:
I added the test for different test. But I didn't know that safe casting is not supported yet in cupy. And because of that, the new test fails for `a` and `values` with different dtypes. 

What should we do in this case for compatibility with numpy?

rushabh-v commented:
IMO, potential options are to make casting `unsafe` OR to restrict different `dtypes` for `a` and `values` and raise an error if they are different.

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +69,61 @@ def _can_memcpy(dst, src):
+        a (cupy.ndarray): Target array.
+        mask (cupy.ndarray): Boolean mask array. It has to be
+            the same shape as `a`.
+        values (cupy.ndarray): Values to put into `a` where `mask` is True.
```
For a frequent case, please support scalar `values`.
```suggestion
        values (cupy.ndarray or scalar): Values to put into `a` where `mask` is True.
```

rushabh-v replied:
Okay, will do that. But in case you haven't noticed my latest comments above, can you please respond to them? @asi1024 
So that I can commit both the changes together. See https://github.com/cupy/cupy/pull/3261#issuecomment-610927701 and https://github.com/cupy/cupy/pull/3261#issuecomment-610931251

asi1024 replied:
How about using `numpy.can_cast`?

rushabh-v replied:
I have already used that in testing. But the tests are failing for even safe casting from numpy.int16 to numpy.int64. It says `casting is not supported yet`.

asi1024 replied:
So, how about using unsafe casting, checking types with `numpy.can_cast` and raising errors if needed?

asi1024 replied:
BTW, I think you can implement it without this typecast. Please avoid it for the performance.

rushabh-v replied:
>BTW, I think you can implement it without this typecast. Please avoid it for the performance.

Okay, so I am removing that casting and the test with different dtypes.

asi1024 replied:
No, don't remove the tests.

rushabh-v replied:
But they will fail on the kernel if the dtypes of `a` and `values` are different.

asi1024 replied:
Please fix to pass tests without typecasting. If you cannot, it's OK to just leave a TODO.

rushabh-v replied:
Can you take a look now?

asi1024 replied:
I mean, please keep the compatibility to NumPy without typecasting.

rushabh-v replied:
okay, then would casting the data inside the Kernel perform better?

rushabh-v replied:
>okay, then would casting the data inside the Kernel perform better? I mean I have seen this somewhere in the library.

I did this. Can you please take a look and tell me whether it is a good way or not?

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +70,69 @@ def _can_memcpy(dst, src):
+
+    else:
+        if not numpy.can_cast(values.dtype, a.dtype):
+            raise TypeError("Cannot cast array data from dtype('float64') to "
+                            "dtype('int64') according to the rule 'safe'")
```
```suggestion
            raise TypeError(
                'Cannot cast array data from {} to {} according to the rule \'safe\''.format(
                    values.dtype, a.dtype)
```

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +70,69 @@ def _can_memcpy(dst, src):
+    if not a.shape == mask.shape:
+        raise ValueError('mask and data must be the same size')
+
+    if cupy.isscalar(values):
+        a[mask] = values
```
Could you add tests for this case?

rushabh-v replied:
I have already added that test.

asi1024 replied:
Thanks!

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +70,69 @@ def _can_memcpy(dst, src):
+        else:
+            if values.ndim > 1:
+                values = values.flatten()
+            _putmask_kernel(mask.astype(numpy.int64),
```
Why is this typecast needed?

rushabh-v replied:
because in the kernel it needed it to be `int` in the if condition. But maybe we can also typecast it there in the kernel directly. I am trying that out.

rushabh-v replied:
oh, when I actually first tried passing `bool` values to the kernel I got an error(I don't remember what but something related to dtypes). And I thought we have to pass ints to the kernels instead of bools. But I just tried again and confirmed that we can pass bools directly also.

asi1024 replied:
Please check `mask.dtype.kind` and raise an error if needed.

asi1024 replied:
Could you add tests for non-bool mask?

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +70,69 @@ def _can_memcpy(dst, src):
+            a[mask] = values[mask]
+
+        else:
+            if values.ndim > 1:
+                values = values.flatten()
```
For the performance.
```suggestion
            values = values.ravel()
```

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +70,69 @@ def _can_memcpy(dst, src):
+    if cupy.isscalar(values):
+        a[mask] = values
+
+    else:
+        if not numpy.can_cast(values.dtype, a.dtype):
```
Why not use `elif`?

rushabh-v commented:
Added the checks for the arguments to be of valid types with the last commit and have made all other requested changes too.

asi1024 commented on cupy/manipulation/basic.py:
```diff
@@ -69,3 +70,74 @@ def _can_memcpy(dst, src):
+        a[mask] = values[mask]
+
+    else:
+        if values.ndim > 1:
```
This condition is needed?
```suggestion
        if values.ndim > 1:
```

rushabh-v replied:
removed it!

rushabh-v commented:
Added support and test for mask of every dtype. NumPy also supports mask of every kind of dtype.

rushabh-v commented:
Could you please take a look @takagi ?

takagi commented:
Yes, sorry for being late.

takagi commented:
CuPy's file structure follows NumPy's reference, where `numpy.putmask` is listed on its indexing routines:
https://numpy.org/doc/stable/reference/routines.indexing.html#inserting-data-into-arrays

Would you move your change into `cupy/indexing/insert.py` ?

rushabh-v commented:
Thanks for the review.
I have made that change. Can you please take a look back?

takagi commented on cupy/indexing/insert.py:
```diff
@@ -66,7 +67,77 @@ def put(a, ind, v, mode='wrap'):
 
 
-# TODO(okuta): Implement putmask
+_putmask_kernel = core._kernel.ElementwiseKernel(
```
Would you remove a private submodule name?
```suggestion
_putmask_kernel = core.ElementwiseKernel(
```

takagi commented on cupy/indexing/insert.py:
```diff
@@ -66,7 +67,77 @@ def put(a, ind, v, mode='wrap'):
+
+def putmask(a, mask, values):
+    """
+    Changes elements of an array inplace, based on conditional mask and
```
```suggestion
    Changes elements of an array inplace, based on a conditional mask and
```

takagi commented on cupy/indexing/insert.py:
```diff
@@ -66,7 +67,77 @@ def put(a, ind, v, mode='wrap'):
+    Sets ``a.flat[n] = values[n]`` for each n where ``mask.flat[n]==True``.
+    If `values` is not the same size as `a` and `mask` then it will repeat.
+
+    Args
```
This colon is necessary: https://google.github.io/styleguide/pyguide.html#doc-function-args
```suggestion
    Args:
```

takagi commented on cupy/indexing/insert.py:
```diff
@@ -66,7 +67,77 @@ def put(a, ind, v, mode='wrap'):
+    """
+
+    if not isinstance(a, cupy.ndarray):
+        raise ValueError('`a` should be of type cupy.ndarray')
```
Would you raise `TypeError` here?

takagi commented on cupy/indexing/insert.py:
```diff
@@ -66,7 +67,77 @@ def put(a, ind, v, mode='wrap'):
+    if not isinstance(a, cupy.ndarray):
+        raise ValueError('`a` should be of type cupy.ndarray')
+    if not isinstance(mask, cupy.ndarray):
+        raise ValueError('`mask` should be of type cupy.ndarray')
```
ditto(TypeError)

takagi commented on cupy/indexing/insert.py:
```diff
@@ -66,7 +67,77 @@ def put(a, ind, v, mode='wrap'):
+    if not isinstance(mask, cupy.ndarray):
+        raise ValueError('`mask` should be of type cupy.ndarray')
+    if not (cupy.isscalar(values) or isinstance(values, cupy.ndarray)):
+        raise ValueError('`values` should be of type cupy.ndarray')
```
ditto(TypeError)

takagi commented:
pfnCI, test this please.

rushabh-v commented:
I have made the changes you requested @takagi

takagi commented:
Thanks, would you add `putmask` to the following file to generate its reference?
https://github.com/rushabh-v/cupy/blob/putmask/docs/source/reference/indexing.rst

takagi commented:
Also, I've found that `cupy.putmask` does not work as `numpy.putmask` in the following case, would you fix it?

```Python
>>> import numpy
>>> import cupy
>>> a = numpy.array([1, 2, 3, 4])
>>> m = numpy.array([0, 1, 2, 3])
>>> numpy.putmask(a, m, 0)
>>> a
array([1, 0, 0, 0])
>>> a = cupy.array([1, 2, 3, 4])
>>> m = cupy.array([0, 1, 2, 3])
>>> cupy.putmask(a, m, 0)
>>> a
array([0, 0, 0, 0])
```

rushabh-v commented:
>Also, I've found that cupy.putmask does not work as numpy.putmask in the following case, would you fix it?

Fixed it and also added a test for that! 
And added `cupy.putmask` to `indexing.rst` too.

takagi commented:
pfnCI, test this please.

rushabh-v commented:
There was a problem in the example in the docstring which was causing the failure in Jenkins build. I have fixed it. It should pass all the checks now.

takagi commented:
pfnCI, test this please.

rushabh-v commented:
Oh, I am sorry. There was a spacing mistake between the elements of the output array in the example. I have fixed it now!

rushabh-v commented:
I am wondering if I can run these checks locally or not!

asi1024 commented:
You can check it locally with `make doctest` in `docs` directory.

takagi commented on tests/cupy_tests/indexing_tests/test_insert.py:
```diff
@@ -115,6 +115,91 @@ def test_put_mode_error(self, dtype):
+class TestPutmaskEqual(unittest.TestCase):
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_allclose()
```
```suggestion
    @testing.numpy_cupy_array_equal()
```

takagi commented on tests/cupy_tests/indexing_tests/test_insert.py:
```diff
@@ -115,6 +115,91 @@ def test_put_mode_error(self, dtype):
+    @testing.numpy_cupy_allclose()
+    def test_putmask(self, xp, dtype):
+        a = testing.shaped_random(self.shape, xp, dtype=dtype, seed=0)
+        mask = testing.shaped_random(self.shape, xp, dtype=bool, seed=1)
```
```suggestion
        mask = testing.shaped_random(self.shape, xp, dtype=numpy.bool_, seed=1)
```

takagi commented on tests/cupy_tests/indexing_tests/test_insert.py:
```diff
@@ -115,6 +115,91 @@ def test_put_mode_error(self, dtype):
+class TestPutmaskNonEqual(unittest.TestCase):
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_allclose()
```
```suggestion
    @testing.numpy_cupy_array_equal()
```

takagi commented on tests/cupy_tests/indexing_tests/test_insert.py:
```diff
@@ -115,6 +115,91 @@ def test_put_mode_error(self, dtype):
+    @testing.numpy_cupy_allclose()
+    def test_putmask(self, xp, dtype):
+        a = testing.shaped_random(self.shape, xp, dtype=dtype, seed=3)
+        mask = testing.shaped_random(self.shape, xp, dtype=bool, seed=4)
```
```suggestion
        mask = testing.shaped_random(self.shape, xp, dtype=numpy.bool_, seed=4)
```

takagi commented on tests/cupy_tests/indexing_tests/test_insert.py:
```diff
@@ -115,6 +115,91 @@ def test_put_mode_error(self, dtype):
+    *testing.product(
+        {'shape': [(0,), (1,), (2, 3), (2, 3, 4)]}))
+@testing.gpu
+class TestPutmaskEqual(unittest.TestCase):
```
How about `TestPutmaskSameShape`?

takagi commented on tests/cupy_tests/indexing_tests/test_insert.py:
```diff
@@ -115,6 +115,91 @@ def test_put_mode_error(self, dtype):
+        {'shape': [(0,), (1,), (2, 3), (2, 3, 4)],
+         'values_shape': [(2,), (3, 1), (5,)]}))
+@testing.gpu
+class TestPutmaskNonEqual(unittest.TestCase):
```
How about `TestPutmaskDifferentShapes`?

takagi commented on tests/cupy_tests/indexing_tests/test_insert.py:
```diff
@@ -115,6 +115,91 @@ def test_put_mode_error(self, dtype):
+        shape = (2, 3)
+        for xp in (numpy, cupy):
+            a = testing.shaped_random(shape, xp, dtype=a_dtype)
+            mask = testing.shaped_random(shape, xp, dtype=bool)
```
```suggestion
            mask = testing.shaped_random(shape, xp, dtype=numpy.bool_)
```

takagi commented:
pfnCI, test this please.

takagi commented:
Thanks!
